mm2src/coins/z_coin.rs

```diff
@@ -380,6 +429,240 @@ impl ZCoin {
         sync_guard.respawn_guard.watch_for_tx(tx.txid());
         Ok(tx)
     }
+
+    async fn tx_history_from_sql(
+        &self,
+        limit: usize,
+        paging_options: PagingOptionsEnum<i64>,
+    ) -> Result<SqlTxHistoryRes, MmError<SqlTxHistoryError>> {
+        let wallet_db = self.z_fields.light_wallet_db.clone();
+        async_blocking(move || {
+            let db_guard = wallet_db.lock();
+            let conn = db_guard.sql_conn();
+
+            let total_sql = SqlBuilder::select_from(TRANSACTIONS_TABLE)
+                .field("COUNT(id_tx)")
+                .sql()
+                .expect("valid SQL");
+            let total_tx_count = conn.query_row(&total_sql, NO_PARAMS, |row| row.get(0))?;
+
+            let mut sql_builder = SqlBuilder::select_from(name!(TRANSACTIONS_TABLE; "txes"));
+            sql_builder
+                .field("txes.txid")
+                .field("txes.id_tx as internal_id")
+                .field("txes.block as block");
+
+            let offset = match paging_options {
+                PagingOptionsEnum::PageNumber(page) => (page.get() - 1) * limit,
+                PagingOptionsEnum::FromId(id) => {
+                    offset_by_id(conn, &sql_builder, [id], "id_tx", "block DESC, id_tx ASC", "id_tx = ?1")?
+                        .ok_or(SqlTxHistoryError::FromIdDoesNotExist(id))?
+                },
+            };
+
+            let sql = sql_builder
+                .field("blocks.time")
+                .field("COALESCE(rn.received_amount, 0)")
+                .field("COALESCE(sn.sent_amount, 0)")
+                .left()
+                .join("(SELECT tx, SUM(value) as received_amount FROM received_notes GROUP BY tx) as rn")
+                .on("txes.id_tx = rn.tx")
+                // detecting spent amount by "spent" field in received_notes table
+                .join("(SELECT spent, SUM(value) as sent_amount FROM received_notes GROUP BY spent) as sn")
+                .on("txes.id_tx = sn.spent")
+                .join(BLOCKS_TABLE)
+                .on("txes.block = blocks.height")
+                .group_by("internal_id")
+                .order_by("block", true)
+                .order_by("internal_id", false)
+                .offset(offset)
+                .limit(limit)
+                .sql()
+                .expect("valid query");
```

Comment:
Maybe this can be made into a function similar to https://github.com/KomodoPlatform/atomicDEX-API/blob/c52a974f239fe8015a2a4b134f4417096e4dbeec/mm2src/coins/tx_history_storage/sql_tx_history_storage_v2.rs#L277-L283 and other similar functions, to simplify `tx_history_from_sql`.

Reply:
It's just a few lines of code. At least for me personally, it will look even harder to read because `finalize` can mean literally anything 🙂

Reply:
Also, after this change https://github.com/KomodoPlatform/atomicDEX-API/pull/1336#discussion_r920246639, `tx_history_from_sql` body is < 60 lines long with almost zero nesting. Looks simple enough already 🙂